packages/xdl/src/detach/AssetBundle.ts

```diff
@@ -0,0 +1,49 @@
+import _ from 'lodash';
+import fs from 'fs-extra';
+import path from 'path';
+import url from 'url';
+import pMap from 'p-map';
+import pRetry from 'p-retry';
+
+import StandaloneContext from './StandaloneContext';
+import { saveUrlToPathAsync } from './ExponentTools';
+
+const EXPO_DOMAINS = ['expo.io', 'exp.host', 'expo.test', 'localhost'];
+const ASSETS_DIR_DEFAULT_URL = 'https://d1wp6m56sqw74a.cloudfront.net/~assets';
+
+type UrlResolver = (hash: string) => string;
+
+export async function bundleAsync(context: any, assets: string[], dest: string) {
+  if (!assets) {
+    return;
+  }
+
+  await fs.ensureDir(dest);
+
+  const urlResolver = createAssetsUrlResolver(context);
+  await pMap(assets, asset => downloadAssetAsync(urlResolver, dest, asset), { concurrency: 5 });
+}
+
+async function downloadAssetAsync(urlResolver: UrlResolver, dest: string, asset: string) {
+  const extensionIndex = asset.lastIndexOf('.');
+  const prefixLength = 'asset_'.length;
+  const hash =
+    extensionIndex >= 0
+      ? asset.substring(prefixLength, extensionIndex)
+      : asset.substring(prefixLength);
+  await pRetry(() => saveUrlToPathAsync(urlResolver(hash), path.join(dest, asset)), { retries: 3 });
+}
+
+function createAssetsUrlResolver(context: StandaloneContext): UrlResolver {
+  let assetsDirUrl = ASSETS_DIR_DEFAULT_URL;
+  if (context && context.published && context.published.url) {
+    const { assetUrlOverride = './assets' } = context.config;
+    const publishedUrl = context.published.url;
+    const hostname = url.parse(publishedUrl).hostname || '';
```

Comment:
Does it make sense to have an empty-string hostname? If the type of `url.parse().hostaname` includes falsy values, maybe we should break early.

Reply:
This is a really good question. I'm not sure if it's possible for `context.published.url` to not have a hostname, but since this would have previously caused a crash ("TypeError: Cannot read property 'split' of null"), I decided to keep in line with that behavior and added a null check that throws if `hostname` is `null`.